### PR TITLE
Expose tldSkipPatterns as property

### PR DESCRIPTION
**Pull request by ptahchiev**
Allow users to customize the tomcat's tld skip patterns via a configuration property.

Related to spring-projects/spring-boot#5010

### CODE REVIEW

**Comment by ptahchiev** [comment]
To justify this pull requested I'd like to mention that my project has at least 5 dependencies with a `Class-Path` entry in their `MANIFEST.MF`. As a result during startup I get about 20 warnings of this kind:

````
2016-08-09 11:36:23,078 org.apache.tomcat.util.scan.StandardJarScanner [localhost-startStop-1] WARN : Failed to scan [file:/home/petar/.m2/repository/org/geotools/gt-metadata/15.0/jts-1.13.jar] from classloader hierarchy
java.io.FileNotFoundException: /home/petar/.m2/repository/org/geotools/gt-metadata/15.0/jts-1.13.jar (No such file or directory)
	at java.util.zip.ZipFile.open(Native Method)
	at java.util.zip.ZipFile.<init>(ZipFile.java:219)
	at java.util.zip.ZipFile.<init>(ZipFile.java:149)
	at java.util.jar.JarFile.<init>(JarFile.java:166)
	at java.util.jar.JarFile.<init>(JarFile.java:130)
	at org.apache.tomcat.util.scan.JarFileUrlJar.<init>(JarFileUrlJar.java:60)
	at org.apache.tomcat.util.scan.JarFactory.newInstance(JarFactory.java:43)
	at org.apache.tomcat.util.scan.StandardJarScanner.process(StandardJarScanner.java:326)
	at org.apache.tomcat.util.scan.StandardJarScanner.scan(StandardJarScanner.java:276)
	at org.apache.jasper.servlet.TldScanner.scanJars(TldScanner.java:262)
	at org.apache.jasper.servlet.TldScanner.scan(TldScanner.java:104)
	at org.apache.jasper.servlet.JasperInitializer.onStartup(JasperInitializer.java:103)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5167)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:152)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1403)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1393)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
````

So for me this pull-request is not a nice-to-have, but is really a must.

**Comment by wilkinsona** [comment]
I don't understand why this is a must. You can already configure this programatically using `TomcatEmbeddedServletContainerFactory.setTldSkip(String)`. Making it configurable via a property seems like a nice-to-have to me.

**Comment by ptahchiev** [comment]
Hi @wilkinsona you are absolutely right. The thing is I'm building a product and would love users to customize this with less coding, but that's actually my problem, not spring-boot's :)

**Comment by ptahchiev** [comment]
Another benefit of having this configured from `application.properties` is that you can have profile specific properties. I just played with the oracle jdbc driver and I got warnings for these files:
````
/home/petar/.m2/repository/com/oracle/jdbc/orai18n/12.1.0.2/orai18n-tools.jar
/home/petar/.m2/repository/com/oracle/jdbc/orai18n/12.1.0.2/orai18n-lcsd.jar
/home/petar/.m2/repository/com/oracle/jdbc/orai18n/12.1.0.2/orai18n-servlet.jar
....
````
Now I have a profile `application-oracle.properties` so if we could configure this property from there I can exclude those jars only for the oracle profile instead of writing in the java code.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java [hunk 1]
```diff
@@ -657,6 +657,8 @@ public void setMaxAge(Integer maxAge) {
 		 */
 		private Charset uriEncoding;
 
+		private String tldSkipPatterns;
```
Comment (snicoll) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java]: Missing documentation

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java [hunk 2]
```diff
@@ -806,6 +816,10 @@ void customizeTomcat(ServerProperties serverProperties,
 			if (this.redirectContextRoot != null) {
 				customizeRedirectContextRoot(factory, this.redirectContextRoot);
 			}
+
+			if (this.tldSkipPatterns != null) {
+				factory.setTldSkip(this.tldSkipPatterns);
+			}
```
Comment (snicoll) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java]: It'd be nice to have a basic test to assert that setting that property has the desired effect. Also the key must be documented in the appendix.

**Comment by snicoll** [comment]
@ptahchiev I've asked for some changes in the PR and due to the lack of answer I went ahead and implemented it (e2b84ea)

@philwebb @wilkinsona the current behaviour is to replace our defaults with anything that's specified via `setTldSkip`. Yet, those defaults are quite useful. I can see two ways to manage that:

1. Add a `setTldSkip(String pattern, boolean additive)` with the existing method calling it with `false` and we can expose a `server.tomcat.additional-tld-skip` that calls it with `true`
2. Change the semantic of `setTldSkip` to *add* skip patterns to the default ones. The drawback is that there is no way to specify the full list anymore (and the semantic of the setter becomes a bit ugly).

Thoughts?

**Comment by snicoll** [comment]
Actually, let's continue that discussion in #5010

**Pull request closed without merging**